### PR TITLE
WIP: turn on specialization spoofing mechanism in the compiler

### LINKED ISSUE #24676: Allow CodeInfo-emitting generated functions to somewhat spoof inference's work-limiting heuristics
Currently, [Cassette](https://github.com/jrevels/Cassette.jl)'s mechanism for intercepting method calls triggers certain inference heuristics that wouldn't normally be triggered on the underlying calls themselves. This can cause inference to give up on inferring a bit too early in some cases, hindering the performance of the resulting code.

For example: Cassette wraps every downstream function call in its own callable type, which causes inference's recursion limiting heuristic to think "hmm, there's a lot of the same function calling itself here, so I should stop inferring now before this recursive madness gets out of hand."

Here's a two-part solution to this problem that @vtjnash, @keno and I cooked up:

1) Add an extra `method_for_heuristic_purposes` field to `CodeInfo`. When inference runs certain heuristics, it will check if this field is inflated. If so, it can use the method in this field rather than the "actual" method for certain parts of the heuristic calculation. This way, `CodeInfo`-emitting `@generated` functions can spoof inference's work-limiting heuristics by inflating the field with the appropriate method.

2) As it stands, `@generated` functions can't actually leverage the above solution, because the limits that need to be spoofed are computed before generator expansion time. To overcome this hurdle, we can add an `expand_generator_early` flag to `Method` that inference can query before it starts applying certain heuristics. If this flag is set for a `@generated` method call that inference is working on, inference will expand the generator earlier than normal (e.g. pre-`abstract_call_method`), thus making the `method_for_heuristic_purposes` field available to downstream heuristics.

This approach should fix some performances problems for Cassette-like packages by allowing them to tell inference "Hey, don't use my weird interception functor to calculate heuristic work limits; use this underlying method instead."

cc @maleadt

### PR DESCRIPTION
**Pull request by jrevels**
Follow-up to #24852. Intends to close #24676.

The intention of this PR is to rewrite relevant parts of the compiler to respect the `signature_for_specialization_heuristics` `CodeInfo` field. I've started off with some very naive changes to `abstract_call_method` that are probably incorrect...

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -189,16 +189,18 @@ function abstract_call_method(method::Method, @nospecialize(sig), sparams::Simpl
...
         infstate = infstate::InferenceState
-        if method === infstate.linfo.def
-            if infstate.linfo.specTypes == sig
-                # avoid widening when detecting self-recursion
-                # TODO: merge call cycle and return right away
-                topmost = nothing
-                edgecycle = true
-                break
-            end
+        if infstate.linfo.specTypes == sig
```
Comment (jrevels) [inline base/compiler/abstractinterpretation.jl]: Should checks against signatures be replaced in the same manner as checks against methods?

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -189,16 +189,18 @@ function abstract_call_method(method::Method, @nospecialize(sig), sparams::Simpl
...
+            # avoid widening when detecting self-recursion
+            # TODO: merge call cycle and return right away
+            topmost = nothing
+            edgecycle = true
+            break
+        end
+        working_method = method_for_specialization_heuristics(infstate.src, infstate.linfo.def)
+        if checked_method === working_method
```
Comment (jrevels) [inline base/compiler/abstractinterpretation.jl]: In general, should all method checks (that are part of specialization heuristics) be replaced like this? In general, when should the "spoof" method be used vs. the "real" method?

**Comment by Keno** [comment]
@jrevels Maybe pull the rename of the CodeInfo field out of this patch and merge, so it doesn't conflict?

**Comment by jrevels** [comment]
Closing because @vtjnash is going to recreate the work here in a different (hopefully more correct) PR

**Pull request closed without merging**